src/Nethermind/Nethermind.Core/Tasks/AsyncManualResetEventSlim.cs

```diff
@@ -0,0 +1,58 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Nethermind.Core.Threading;
+
+namespace Nethermind.Core.Tasks;
+
+public class AsyncManualResetEventSlim
+{
+    private volatile TaskCompletionSource<bool> _tcs = CreateNewTcs();
+
+    private static TaskCompletionSource<bool> CreateNewTcs() =>
+        new(TaskCreationOptions.RunContinuationsAsynchronously);
+
+    public AsyncManualResetEventSlim(bool initialState = false)
+    {
+        if (initialState)
+        {
+            _tcs.TrySetResult(true);
+        }
+    }
+
+    public bool IsSet => _tcs.Task.IsCompleted;
+
+    public void Set() => _tcs.TrySetResult(true);
+
+    public void Reset()
+    {
+        TaskCompletionSource<bool>? newTcs = null;
+        while (true)
+        {
+            TaskCompletionSource<bool> tcs = _tcs;
+            if (!tcs.Task.IsCompleted)
+                return;
+
+            newTcs ??= CreateNewTcs();
+            if (Interlocked.CompareExchange(ref _tcs, newTcs, tcs) == tcs)
```

Comment:
You interlock here but in `Set` you just read the field with `volatile`. This could result in Reset swaping the underlying `tcs` while `Set` would get the previous and try set result on it. What is the contract here? Who's gonna call Set/Reset?

Reply:
This intends to be replacement for `ManualResetEventSlim` but with async wait